symfit/core/minimizers.py

```diff
@@ -172,43 +193,70 @@ def scipy_constraints(constraints, data):
             sympy.Eq: 'eq', sympy.Ge: 'ineq',
         }
 
-        for key, constraint in enumerate(constraints):
-            # jac = make_jac(c, p)
+        for key, partialed_constraint in enumerate(constraints):
+            constraint_type = partialed_constraint.func.constraint_type
             cons.append({
-                'type': types[constraint.constraint_type],
-                # Assume the lhs is the equation.
-                'fun': lambda p, x, c: c(*(list(x.values()) + list(p)))[0],
-                # Assume the lhs is the equation.
-                'jac': lambda p, x, c: [component(*(list(x.values()) + list(p)))
-                                        for component in
-                                        c.numerical_jacobian[0]],
-                'args': (data, constraint)
+                'type': types[constraint_type],
+                # Takes an nd.array of params and a partialed_constraint, and
+                # evaluates the constraint with these parameters.
+                # Wrap `c` so it is always called via keywords.
+                'fun': lambda p, c: self.list2kwargs(c)(list(p))[0],
+                'args': [partialed_constraint]
             })
+
         cons = tuple(cons)
         return cons
 
-class ScipyGradientMinimize(ScipyMinimize, GradientMinimizer):
-    def __init__(self, *args, **kwargs):
-        super(ScipyGradientMinimize, self).__init__(*args, **kwargs)
-        self.wrapped_jacobian = self.wrap_func(self.wrapped_jacobian)
-
-    def execute(self, **minimize_options):
-        return super(ScipyGradientMinimize, self).execute(jacobian=self.wrapped_jacobian, **minimize_options)
-
-
 class BFGS(ScipyGradientMinimize):
     def execute(self, **minimize_options):
         return super(BFGS, self).execute(method='BFGS', **minimize_options)
 
 
-class SLSQP(ScipyGradientMinimize, ConstrainedMinimizer, BoundedMinimizer):
+class SLSQP(ScipyConstrainedMinimize, GradientMinimizer, BoundedMinimizer):
+    def __init__(self, *args, **kwargs):
+        super(SLSQP, self).__init__(*args, **kwargs)
+        # We have to break DRY because you cannot inherit from both
+        # ScipyConstrainedMinimize and ScipyGradientMinimize. So SLQSP is a
+        # special case. This is the same code as in ScipyGradientMinimize.
+        self.wrapped_jacobian = self.list2kwargs(self.wrapped_jacobian)
```

Comment:
That mean we did *something* wrong in class design. I don't really see where though.
Probably `ScipyMinimize` has to actually do something, and `ScipyGradient` and `ScipyConstraint` need to be purely mix-ins that can't exist on themselves. I suggest we leave it for now.

Reply:
Yeah, that was also my thinking. They should probably be made into pure mixin classes, but since SLSQP is the only one breaking the pattern we can treat it like a special snowflake for now ;).

Reply:
We might want to expand the comment with this. Just so we don't have to come up with this again